hydra-node/src/Hydra/API/Server.hs

```diff
@@ -66,48 +66,63 @@ type ServerCallback tx m = ClientInput tx -> m ()
 type ServerComponent tx m a = ServerCallback tx m -> (Server tx m -> m a) -> m a
 
 withAPIServer ::
+  forall tx.
   (IsChainState tx) =>
   IP ->
   PortNumber ->
   Party ->
-  PersistenceIncremental (ServerOutput tx) IO ->
+  PersistenceIncremental (TimedServerOutput tx) IO ->
   Tracer IO APIServerLog ->
   ServerComponent tx IO ()
 withAPIServer host port party PersistenceIncremental{loadAll, append} tracer callback action = do
   responseChannel <- newBroadcastTChanIO
   h <- loadAll
-  -- NOTE: We will add a 'Greetings' message on each API server start. This is
-  -- important to make sure the latest configured 'party' is reaching the
-  -- client.
   -- NOTE: we need to reverse the list because we store history in a reversed
   -- list in memory but in order on disk
   history <- newTVarIO (reverse h)
 
-  appendToHistory history $ Greetings party
+  -- NOTE: We will add a 'Greetings' message on each API server start. This is
+  -- important to make sure the latest configured 'party' is reaching the
+  -- client.
+  void $ appendToHistory history (Greetings party)
   race_
     (runAPIServer host port tracer history callback responseChannel)
     . action
     $ Server
       { sendOutput = \output -> do
-          appendToHistory history output
+          timedOutput <- appendToHistory history output
           atomically $
-            writeTChan responseChannel output
+            writeTChan responseChannel timedOutput
       }
  where
-  appendToHistory history event = do
-    append event
-    atomically $
-      modifyTVar' history (event :)
+  appendToHistory :: TVar [TimedServerOutput tx] -> ServerOutput tx -> IO (TimedServerOutput tx)
+  appendToHistory history output = do
+    time <- getCurrentTime
+    timedOutput <- atomically $ do
+      seq <- nextSequenceNumber history
+      let timedOutput = TimedServerOutput{output, time, seq}
+      modifyTVar' history (timedOutput :)
+      pure timedOutput
+    append timedOutput
+    pure timedOutput
+
+nextSequenceNumber :: TVar [TimedServerOutput tx] -> STM.STM Natural
```

Comment:
can we change the function definition to:
```hs
nextSequenceNumber :: TimedServerOutput tx -> Natural
```

wdyt?

Reply:
Then we need another function which goes `TVar [TimedServerOutput tx] -> STM (TimedServerOutput tx)` + the one you suggest.

Why?

Reply:
but i was thinking to call it like this:

```hs
seq <- nextSequenceNumber <$> readTVar history
```